x/evm/keeper/msg_server.go

```diff
@@ -153,9 +163,64 @@ func (k *Keeper) UpdateParams(goCtx context.Context, req *types.MsgUpdateParams)
 	}
 
 	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	if err := k.updatePrecompiles(ctx, req.Params.EnabledPrecompiles); err != nil {
+		return nil, err
+	}
+
 	if err := k.SetParams(ctx, req.Params); err != nil {
 		return nil, err
 	}
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+func (k *Keeper) updatePrecompiles(ctx sdk.Context, newEnabledPrecompiles []string) error {
+	oldParams := k.GetParams(ctx)
+	oldEnabledPrecompiles := oldParams.GetEnabledPrecompiles()
+
+	newEnabledPrecompilesMap := make(map[common.Address]struct{}, len(newEnabledPrecompiles))
+
+	// initialize newly enabled precompiles
+	for _, hexAddr := range newEnabledPrecompiles {
+		addr := common.HexToAddress(hexAddr)
+		newEnabledPrecompilesMap[addr] = struct{}{}
+
+		account := k.GetAccount(ctx, addr)
+		if account != nil && account.Nonce == PrecompileNonce && bytes.Equal(account.CodeHash, PrecompileCodeHash[:]) {
+			// skip initialization if account is already initialized
+			continue
+		}
+
+		// Set the nonce of the precompile's address (as is done when a contract is created) to ensure
+		// that it is marked as non-empty and will not be cleaned up when the statedb is finalized.
+		err := k.SetAccount(ctx, addr, statedb.Account{
+			Nonce:    PrecompileNonce,
+			Balance:  big.NewInt(0),
+			CodeHash: PrecompileCodeHash[:],
+		})
+		if err != nil {
+			return err
+		}
+
+		// Set the code of the precompile's address to a non-zero length byte slice to ensure that the precompile
+		// can be called from within Solidity contracts. Solidity adds a check before invoking a contract to ensure
+		// that it does not attempt to invoke a non-existent contract.
+		k.SetCode(ctx, PrecompileCodeHash[:], PrecompileCode)
```

Comment:
SetCode could be done just once since it's the same for all contracts

Reply:
make sense, fixed